tests/foreman/api/test_puppetmodule.py

```diff
@@ -50,11 +40,11 @@ def test_positive_search_no_results(self):
 
         :CaseImportance: Critical
         """
-        query = {'repository_id': self.repository.id}
-        self.assertEqual(len(entities.PuppetModule().search(query=query)), 0)
+        query = {'repository_id': make_repo.id}
+        assert len(entities.PuppetModule().search(query=query)) == 0
```

Comment:
You made `make_repo` fixture on class level, which means that all tests within one class depending on this fixture will get the same instance. In another test, you upload puppet module to repository. Which means that if it happens that `test_positive_search_single_result` is executed first, and `test_positive_search_no_results` is executed second, this assertion will fail - in other words, test result depend on tests execution order. That is usually discouraged.

You need to change scope of your fixture, or create new repository directly in one of tests.

Reply:
Yes, that's why I changed index in `make_cv.version[0].id` from `0` to `-1`, so that the latest published CV version with appropriate puppet modules count is queried, regardless the execution order (cv.publish() is called in each test).

However, you are right this not so straightforward and the only plus is slightly shorter execution time. So I've pulled the fixtures out of classes and refactored to use different contentviews for each test (as was the original way with setUp function).

All tests are currently green, I'll paste the results after all comments are addressed.

Reply:
@vsedmik I don't think @mirekdlugosz concern is fully addressed here.

`make_product_repo` is module scoped, so one product and one repository will be created for all of the test functions in the module.

`test_positive_search_no_results` uses this single repository entity to assert the length of the PuppetModule search is zero

`test_positive_search_single_result` uses this single repository entity to call `upload_content`, and then does the same PuppetModule search to assert the query result length is 1.  The test does not remove this uploaded content from the repository.

We don't guarantee test execution order. Running against the same Satellite, if `test_positive_search_single_result` runs first it will upload content to the singular module-scoped Repository.  When `test_positive_search_no_results` runs, it will have content uploaded to the repository, and its length check will fail as now the length of the search result will be 1.

While the product entity can be module scoped, I think that the repository entity should be function scoped, to deal with the above behavior and retain the behavior of the original unittest `setUpClass` and `setUp` method scopes.